### ISSUE TITLE
Improve package loading performance

### ISSUE THREAD

**Issue by abdonrd** [sindresorhus/atom-editorconfig#66]
I use [Timecop](https://github.com/atom/timecop) to see the information about where time is spent while Atom loads and find this problem:

![screen shot 2016-02-17 at 10 39 08](https://cloud.githubusercontent.com/assets/1007051/13105353/11377d4c-d563-11e5-8600-e78f72f3627d.png)

**Comment by sindresorhus** [comment]
Help welcome :)

**Comment by florianb** [comment]
I'm taking this issue. :racehorse:

**Comment by florianb** [comment]
Tests are implemented currently - unfortunately the measurements are significantly lower than in a production-environment. The test currently fails with a loadtime of `9-14ms` depending on the run.

Unless we're able to test the real behavior i hope that our measures to tune editorconfig which then will pass the test change the production behavior as expected, too.

**Comment by florianb** [comment]
I created a first draft of the performance-optimization (https://github.com/florianb/atom-editorconfig/tree/iss66). Unfortunately i'm not satisfied with the results.

Strangely, when playing around with the different components i found out that having Xo as devDependency the whole load/activation-process slows down. I currently have no idea why this happens - probably @sindresorhus has?

However - i will do some further investigations. :eyeglasses:

**Comment by sindresorhus** [comment]
Do you happen to have `atom-linter-xo` installed?

You can do more advanced perf profiling in Dev Tools.

**Comment by florianb** [comment]
Yes, i use `atom-linter-xo`. I haven't used the Chrome-Dev-Profiler.. but i guess that's just something i have to learn now.. :dancers:

**Comment by sindresorhus** [comment]
The framegraph is your friend https://addyosmani.com/blog/devtools-flame-charts/

**Comment by florianb** [comment]
Great - thanks! :clap:

**Comment by florianb** [comment]
Okay - first riddle solved, calling `workspace.observeTextEditors` consumes 92% of the activation-time.

I will try to find out on which event we could try to safely rely on, in order to manage not missing any spawning editors.

**Comment by florianb** [comment]
Okay - i did some further tests. Relying on other events will break the functionality.

There are two facts we have to accept. At first, due to the lazy-load-mechanisms of Atom, the requirements for `observerTextEditors` will get into our Timecop-account as long as editorconfig is accessing that event-subscription first.

Secondly, Timecop-measurements differ from machine to machine, from environment to environment. This makes the whole issue untestable. I therefore put a **upper load limit** of 50ms into action. This should prevent future changes which cause long load-times.

**Comment by Arcanemagus** [comment]
Btw @florianb two things about load times:

* Atom by default doesn't install `devDependencies` when a package is installed regularly, so `xo` being listed shouldn't affect a _normal_ installation
* Atom run through `--dev` mode loads all dependencies regardless of if the package's code actively uses them.
* Atom package load time is dependent on the list of dependencies installed, regardless of when and/or if they are being loaded. You can see some hard data on that [here](https://docs.google.com/spreadsheets/d/1XkYcoIk9bCMo-WtAttYXrgsVilHfTmvAAPT_8Kei0Nc/edit#gid=652278334), note that `apm install` there is from running `apm install` within the package directory (not installing from within Atom or running `apm installl package-name`). Also the difference between `apm install` and `npm install` is due to that test having been run with `npm@3` which flattens the directory structure as much as possible, while `apm` is based on `npm@2` which uses a hierarchical structure that apparently takes _far_ longer to parse.

tl;dr: Having `xo` there shouldn't affect production installs :stuck_out_tongue:
